lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -914,6 +936,106 @@ makeChange minCoinValueFor requiredCost mExtraCoinSource inputBundles outputBund
+--
+assignCoinsToChangeMaps
+    :: Coin
+    -- ^ The total quantity of ada available, including any extra source of ada.
+    -> (TokenMap -> Coin)
+    -- ^ A function to calculate the minimum required ada quantity for any
+    -- token map.
+    -> NonEmpty (TokenMap, Coin)
+    -- ^ A list of pre-computed asset change maps paired with original output
+    -- coins, sorted into an order that ensures all empty token maps are at the
+    -- start of the list.
+    -> Maybe [TokenBundle]
+    -- ^ Resulting change bundles, or 'Nothing' if there was not enough ada
+    -- available to assign a minimum ada quantity to all non-empty token maps.
+assignCoinsToChangeMaps adaAvailable minCoinFor pairsAtStart =
+    loop adaRequiredAtStart pairsAtStart
+  where
+    loop !adaRequired !pairsNonEmpty = case pairsNonEmpty of
+
+        pair :| pairs | adaAvailable >= adaRequired ->
+            -- We have enough ada available to pay for the minimum required
+            -- amount of every asset map that remains in our list:
+            let assetMapsRemaining = fst <$> (pair :| pairs) in
+            let bundlesForAssetsWithMinimumCoins =
+                    assignMinimumCoin minCoinFor <$> assetMapsRemaining in
+            -- Calculate the amount of ada that remains after assigning the
+            -- minimum amount to each map. This should be safe, as we have
+            -- already determined that we have enough ada available:
+            let adaRemaining = adaAvailable `Coin.distance` adaRequired in
+            -- Partition any remaining ada according to the weighted
+            -- distribution of output coins that remain in our list:
+            let outputCoinsRemaining = snd <$> (pair :| pairs) in
+            let bundlesForOutputCoins = TokenBundle.fromCoin <$>
+                    makeChangeForCoin outputCoinsRemaining adaRemaining in
+            -- Finally, combine the minimal coin asset bundles with the
+            -- bundles obtained by partitioning the remaining ada amount:
+            Just $ NE.toList $ NE.zipWith (<>)
+                bundlesForAssetsWithMinimumCoins
+                bundlesForOutputCoins
+
+        (m, _) :| (p : ps) | TokenMap.isEmpty m && adaAvailable < adaRequired ->
+            -- We don't have enough ada available to pay for the minimum
+            -- required amount of every asset map, but we do have an empty
+            -- asset map that is safe to drop. This will reduce the amount of
+            -- ada required by a small amount:
+            let adaRequired' = adaRequired `Coin.distance` minCoinFor m in
+            loop adaRequired' (p :| ps)
+
+        (m, _) :| [] | TokenMap.isEmpty m && adaAvailable >= adaRequired ->
+            -- We didn't have any non-ada assets at all in our change. We can
+            -- simply return the available ada amount as a singleton change
+            -- bundle:
+            Just [TokenBundle.fromCoin adaAvailable]
```

Comment:
Isn't that branch a special case of / redundant with the first branch? The first branch will always match before this one, and I'd indeed expect the result to be the same.

Reply:
I think you're right. I've fixed this in 427e208bfddbe07794927f2e3855761b530196cf.